src/printer_tree.c

```diff
@@ -0,0 +1,3107 @@
+ * @param[in,out] tc is the tree context.
+ */
+static void
+trm_print_groupings(struct trt_printer_ctx *pc, struct trt_tree_ctx *tc)
+{
+    ly_bool once = 1;
+
+    for (struct trt_keyword_stmt ks = pc->fp.modify.next_grouping(tc);
+            !trp_keyword_stmt_is_empty(ks);
+            ks = pc->fp.modify.next_grouping(tc)) {
+        if (once) {
+            ly_print_(pc->out, "\n");
+            ly_print_(pc->out, "\n");
+            once = 0;
+        } else {
+            ly_print_(pc->out, "\n");
+        }
+        trm_print_body_section(ks, pc, tc);
+    }
+}
+
+/**
+ * @brief For all yang-data sections: print 'yang-data' keyword and all its nodes.
+ * @param[in] pc contains mainly functions for printing.
+ * @param[in,out] tc is the tree context.
+ */
+static void
+trm_print_yang_data(struct trt_printer_ctx *pc, struct trt_tree_ctx *tc)
+{
+    ly_bool once = 1;
+
+    for (struct trt_keyword_stmt ks = pc->fp.modify.next_yang_data(tc);
+            !trp_keyword_stmt_is_empty(ks);
+            ks = pc->fp.modify.next_yang_data(tc)) {
+        if (once) {
+            ly_print_(pc->out, "\n");
+            ly_print_(pc->out, "\n");
+            once = 0;
+        } else {
+            ly_print_(pc->out, "\n");
+        }
+        trm_print_body_section(ks, pc, tc);
+    }
+}
+
+/**
+ * @brief Print sections module, augment, rpcs, notifications, grouping, yang-data.
+ * @param[in] pc contains mainly functions for printing.
+ * @param[in,out] tc is the tree context.
+ */
+static void
+trm_print_sections(struct trt_printer_ctx *pc, struct trt_tree_ctx *tc)
+{
+    trm_print_module_section(pc, tc);
+
+    trm_print_augmentations(pc, tc);
+
+    trm_print_rpcs(pc, tc);
+
+    trm_print_notifications(pc, tc);
+
+    trm_print_groupings(pc, tc);
+
+    trm_print_yang_data(pc, tc);
+
+    ly_print_(pc->out, "\n");
+}
+
+/**
+ * @brief Get default settings for trt_printer_ctx.
+ *
+ * Get trt_printer_ctx containing all structure items correctly defined except for trt_printer_opts and max_line_length,
+ * which are parameters of the printer tree module.
+ *
+ * @param[in,out] out is output handler.
+ * @param[in] max_line_length is the maximum line length limit that should not be exceeded.
+ */
+static struct trt_printer_ctx
+trm_default_printer_ctx(struct ly_out *out, size_t max_line_length)
+{
+    struct trt_fp_modify_ctx modify_fp = {
+        tro_modi_parent,
+        tro_modi_first_sibling,
+        tro_modi_next_sibling,
+        tro_modi_next_child,
+        tro_modi_next_augment,
+        tro_modi_get_rpcs,
+        tro_modi_get_notifications,
+        tro_modi_next_grouping,
+        tro_modi_next_yang_data
+    };
+
+    struct trt_fp_read read_fp = {
+        tro_read_module_name,
+        tro_read_node,
+        tro_read_if_sibling_exists
+    };
+
+    struct trt_fp_print print_fp = {
+        tro_print_features_names,
+        tro_print_keys
+    };
+
+    return (struct trt_printer_ctx) {
```

Comment:
Connected to a comment before, why not declare a static variable of this type, fill it properly, and return it? Better yet, set return type to `void` and make `struct trt_printer_ctx *` a parameter that you simply fill correctly. That is how we usually write dynamic initializers, avoids allocating the whole structure as return value and then the copy into the local structure.